### PR TITLE
[C++][Pistache] Fix optional error and wrong function signatures

### PR DESCRIPTION
**Pull request by stkrwork**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

(details of the change, additional tests that have been done, reference to the issue for tracking, etc)

#257 #258

### CODE REVIEW

**Comment by stkrwork** [comment]
@ravinikam @fvarose @etherealjoy @martindelille

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java [hunk 1]
```diff
@@ -327,29 +327,16 @@ public String toDefaultValue(Schema p) {
             ArraySchema ap = (ArraySchema) p;
             String inner = getSchemaType(ap.getItems());
             if (!languageSpecificPrimitives.contains(inner)) {
-                inner = "std::shared_ptr<" + inner + ">";
+                inner = inner;
```
[the remark this answers is not in the document]
Comment (stkrwork) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java]: oh, seems i forgot to remove it, when removing shared ptr

File: samples/server/petstore/pistache-server/model/Pet.cpp [hunk 2]
```diff
@@ -155,11 +155,11 @@ void Pet::unsetId()
 {
     m_IdIsSet = false;
 }
-std::shared_ptr<Category> Pet::getCategory() const
+Category Pet::getCategory() const
 {
     return m_Category;
 }
-void Pet::setCategory(std::shared_ptr<Category> value)
+void Pet::setCategory(Category value)
```
[the remark this answers is not in the document]
Comment (stkrwork) [reply samples/server/petstore/pistache-server/model/Pet.cpp]: i'll add it

File: samples/server/petstore/pistache-server/model/Pet.h [hunk 3]
```diff
@@ -62,8 +62,8 @@ class  Pet
     /// <summary>
     /// 
     /// </summary>
-    std::shared_ptr<Category> getCategory() const;
-    void setCategory(std::shared_ptr<Category> value);
+    Category getCategory() const;
+    void setCategory(Category value);
```
[the remark this answers is not in the document]
Comment (stkrwork) [reply samples/server/petstore/pistache-server/model/Pet.h]: i'll add it

**Comment by etherealjoy** [comment]
@stkrwork 
Passing complex types by reference makes sense since it avoid copy construction. But Primitive types benefit from passing by value instead of by reference.
Does it makes sense to do this also here. I mean for all purposes the functional behavior is the same.

**Comment by MartinDelille** [comment]
I don’t know especially what Pistache is but passing the value by **const** reference, even for complex type, should be generalized to all cpp generators. What do you think?

**Comment by stkrwork** [comment]
i was considering const&, but in my initial thought i was thinking of doing it for every signature, but this approach is better.

**Comment by MartinDelille** [comment]
Each time I consider using const attributes for my methods, it leads to big and healthy refactoring! 😄

**Comment by MartinDelille** [comment]
And for Qt generator too! I create the issue.

**Comment by etherealjoy** [comment]
@stkrwork 
I have a question, now the complex types have to be const before passing as reference to the api right?

**Comment by stkrwork** [comment]
It just states that the complex type wont be modified. they dont have to be const outside of the function, they are just dealt as constants in the function

**Comment by stkrwork** [comment]
@CyrilleBenard does that fix your issues #257 and #258

**Comment by CyrilleBenard** [comment]
@stkrwork I'm sorry, I still do not try your PR 'cause I evaluate the openapi-generator during my time job. I will do it during the next week. Thx for your reactivity

**Comment by CyrilleBenard** [comment]
One regression seems to occurred, see [issue #257 - part 2 of my last comment](https://github.com/OpenAPITools/openapi-generator/issues/257) relative to the use of _static_ + _const_

**Comment by stkrwork** [comment]
i added your request

**Comment by wing328** [comment]
If no further question/feedback on this PR, I'll merge it into master in a few hours.

cc @ravinikam @fvarose @etherealjoy @MartinDelille

**Comment by CyrilleBenard** [comment]
Optional error has been fixed but I as far as I know, the function signature is still buggy. May be I misunderstand what's behind "wrong function signatures" (?).
In my opinion, it refers to [#258](https://github.com/OpenAPITools/openapi-generator/issues/258) and nothing has been fixed :/

**Comment by stkrwork** [comment]
well, it might take some time, because i don't have much time at the moment to implement the ideas i posted in chat

**Comment by CyrilleBenard** [comment]
@stkrwork yes sure :+1: 
I just reply to @wing328's question :)
Hopping the C++ server stub will compile again as soon as possible, maybe next week ? :dancer:

**Comment by stkrwork** [comment]
otherwise you can also pick up where i left off.

**Comment by wing328** [comment]
I'll put this on hold and will only merge this after I get the sign-off from you guys. Please take your time.

**Comment by Schrigga** [comment]
@stkrwork I forked your branch with the changes you have already done. I changed some things, but I stucked compiling the UserApi.cpp when the users should be initialized using an array or an list. 
Currently I'm wondering if there is a possibility in the mustache files to distinguish whether there are non primitive type and are arrays. Do you have any knowledge about this?

**Comment by wing328** [comment]
@Schrigga what about using the following?
```
{{#isListContainer}}
{{^isPrimitiveType}}
...
{{/isPrimitiveType}}
{{/isListContainer}}
```

**Pull request merged**